python_modules/dagster/dagster/cli/dynamic_loader.py

```diff
@@ -208,25 +208,47 @@ def create_pipeline_loading_mode_data(info):
         raise InvalidPipelineLoadingComboError()
 
 
-DynamicObject = namedtuple('DynamicObject', 'module fn module_name fn_name object')
+class DynamicObject:
+    def __init__(self, module, module_name, fn_name):
+        self.module = module
+        self.module_name = module_name
+        self.fn_name = fn_name
+        self.object = None
+        self.coerce_to_repo = False
+        self.loaded = False
+
+    def eval(self):
+        if self.loaded:
+            module = importlib.reload(self.module)
+        self.loaded = True
+
+        fn = getattr(self.module, self.fn_name)
+        check.is_callable(fn)
+        self.object = fn()
+
+        if self.coerce_to_repo and not isinstance(self.object, RepositoryDefinition):
```

Comment:
Let's keep the notion of DynamicObject unaware of application types.

Reply:
Hey—this is a great idea, but I think doing this in a really clean way might require pretty extensive changes because the various DynamicObject factory methods (ala `load_repository_object_from_target_info`) don't actually /load/ the object anymore and can't verify the result is a Repository. (The loading is done in the RepositoryContainer, where errors can be caught.)

I've pushed a commit that adds a "post-load hook" to the DynamicObject so that a dynamic object factory method can specify coercion behavior (via a post-load transformation) without having to load the object. We could also create DynamicObject subclasses (eg DynamicRepositoryObject) but that seems like overkill.

I think a nicer solution would be to make the entire dynamic_object file and it's factory methods type-agnostic and put the pipeline->repository coercion, `RepositoryTargetMode` decision making, `RepositoryTargetInfo`-specific stuff, etc. in `RepositoryContainer`. I think that's a fairly large undertaking though given the tests interface with the DynamicObject's `load_repository_object_from_target_info`  method.